### PR TITLE
Refesh development docs.

### LINKED ISSUE #1122: Split up C unit tests
The C unit tests are all currently in one file. Split this up into (say) three files, ``test_likelihood``, ``test_ancestry``, and ``test_mutations``. We can copy some of the code from tskit's [test_main.c](https://github.com/tskit-dev/tskit/blob/master/c/tests/testlib.c) to avoid repeating all the CUnit boilerplate in each file.

This is a good introduction to the C code (see also #1211 and #1084)

### LINKED ISSUE #1208: add clang-format to developer docs
There currently isn't guidance on how to format the C side of msprime, even though one of the PR checks is about formatting.

If you just mindlessly `clang-format` all the C files you get some changes to `avl.c` and `mutgen.c` that are currently in `main`.

### PR DESCRIPTION
**Pull request by jeromekelleher**
- Update the makefile to be more obvious and add clang-format target
- Make it more difficult to mess up with clang-format by adding
  format-off in files we don't format.
- Fix compile bug that somehow slipped through the net.
- Make a pass through dev docs to refresh.

Closes #1122
Closes #1208

### CODE REVIEW

**Comment by jeromekelleher** [comment]
@mergifyio rebase

File: lib/util.c [hunk 1]
```diff
@@ -265,6 +265,7 @@ msp_strerror_internal(int err)
             break;
         case MSP_ERR_UNKNOWN_TIME_NOT_SUPPORTED:
             ret = "Kept mutations must have known mutation times.";
+            break;
```
Comment (benjeffery) [inline lib/util.c:268]: Oooo, nice catch. Starting to think most tests for exceptions should check the message too!

**Review by benjeffery — APPROVED** [review]
Great, we should add a similar make target to tskit.

File: lib/util.c (hunk 1 above)
Comment (jeromekelleher) [reply lib/util.c:268]: The compiler spotted it - I don't know how it slipped through to be honest. It must have happened at some point when I fixed a merge conflict, but I don't know how CI passed afterwards. :shrug:

Comment (benjeffery) [reply lib/util.c:268]: @jeromekelleher (do I need to ping for comments like these?)

The error was introduced by https://github.com/tskit-dev/msprime/commit/f18650fdc2931bc88eaad334a4188f077ae83298 
This commit (and the ones after) passed all checks, however our CI is on gcc 6 and  `-Wimplicit-fallthrough` was added in gcc 7.

Comment (jeromekelleher) [reply lib/util.c:268]: Aha! Thanks for the detective work. This is unlikely to skip through again, so let's not worry about it.

**Comment by jeromekelleher** [comment]
@Mergifyio update

**Comment by jeromekelleher** [comment]
Hello, mergify :wave:

**Comment by jeromekelleher** [comment]
Ah, mergify seems to have woken up now I've manually rebased. Maybe it got confused by the base branch switch.

**Pull request merged**